File: Timer.py
from typing import Any, Callable, TextIO
import math
from time import perf_counter


def timer_decorator(funk: Callable) -> Callable:
    def wrapper(*args, **kwargs) -> Any:
        start = perf_counter()
        funk(*args, **kwargs)
        end = perf_counter()
        print(f"Done! It took: {end - start : .3f}sec")
    return wrapper


class Timer:
    class _Time:
        def __init__(self, hours: int, minutes: int, seconds: int, millisecond: int) -> None:
            self.hours = hours
            self.minutes = minutes
            self.seconds = seconds
            self.millisecond = millisecond
        
        def __str__(self) -> str:
            return f"{self.hours:0>2}:{self.minutes:0>2}:{self.seconds:0>2},{self.millisecond:0>3}"

    def __init__(self, source: str, file: str, binaryMode: bool = False) -> None:
        if binaryMode:
            self._source = open(source, mode="rb")
            self._file = open(file, mode="wb")
        else:
            self._source = open(source, mode="r", encoding='utf-8')
            self._file = open(file, mode="w", encoding='utf-8')

    @property
    def source(self) -> TextIO:
        return self._source

    @property
    def file(self) -> TextIO:
        return self._file

    def __enter__(self) -> TextIO:
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        if self.source:
            self.source.close()
        if self.file:
            self.file.close()
    
    def pars_time(self, old_time: str) -> _Time:
        hours = int(old_time[0:2])
        minutes = int(old_time[3:5])
        seconds = int(old_time[6:8])
        millisecond = int(old_time[9:12])
        
        return self._Time(hours, minutes, seconds, millisecond)

    def shift(self, time: _Time, shift_by: int) -> _Time:
        shift_time = time.hours * 3_600_000 \
                     + time.minutes * 60_000 \
                     + time.seconds * 1_000 \
                     + time.millisecond \
                     - shift_by
        h = math.floor(shift_time / 3_600_000)
        m = math.floor(shift_time % 3_600_000 / 60_000)
        s = math.floor(shift_time % 60_000 / 1_000)
        ms = math.floor(shift_time % 1_000)
        new_time = self._Time(h, m, s, ms)
        return new_time

    @staticmethod
    def print_time(time: _Time) -> str:
        return str(time)


@timer_decorator
def shift_time_1(file: str, new_file: str, shift_by: int) -> None:
    shift_by *= 1000
    
    with Timer(file, new_file) as shifter:
        text_number = 1
        line_number = 0

        while line := shifter.source.readline():
            if str(text_number) == line.strip() or text_number == 1:
                hour = shifter.source.readline()

                shifted_time = "{} --> {}\n"

                parsed_1st_part = shifter.pars_time(hour[ 0 : hour.find(" --> ") ])
                shifted_parsed_1st_parth = shifter.shift(parsed_1st_part, shift_by)
                shifted_1st_part = shifter.print_time(shifted_parsed_1st_parth)

                parsed_2nd_part = shifter.pars_time(hour[ hour.find(" --> ") + len(" --> ") : -1])
                shifted_parsed_2nd_parth = shifter.shift(parsed_2nd_part, shift_by)
                shifted_2nd_part = shifter.print_time(shifted_parsed_2nd_parth)

                shifted_time = shifted_time.format(shifted_1st_part, shifted_2nd_part)
                shifter.file.write(f"{text_number}\n")
                shifter.file.write(shifted_time)

                text_number += 1

            else:
                shifter.file.write(line)

            line_number += 1


@timer_decorator
def shift_time_2(file: str, new_file: str, shift: int) -> None:
    # f = open(file, "rb")
    # print(f.readline()[2:13])
    # print(b'{' == b'{')
    # f.close()
    brace = ord('}')

    with Timer(file, new_file, binaryMode=True) as new:
        while line := new.source.readline().strip():

            k = 1
            while line[k] != brace:
                k += 1
            time_1st_part = int(line[1:k]) - shift

            k += 2
            k_2 = k
            while line[k] != brace:
                k += 1
            time_2nd_part = int(line[k_2:k]) - shift
            
            new.file.write(b'{%d}{%d}%s\n' % (time_1st_part, time_2nd_part, line[k+1:]))

File: Timer.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <ctype.h>

struct TimeParser_t
{
    int32_t hours, minutes, seconds, milliseconds;
};
typedef struct TimeParser_t TimeParser_t;

struct Timer_t
{
    char* source;
    char* shiftedSubtitles;
};
typedef struct Timer_t Timer_t;

uint32_t getCollapsedTime(TimeParser_t* parser)
{
    return parser->hours * 3600000
         + parser->minutes * 60000
         + parser->seconds * 1000
         + parser->milliseconds;
}

TimeParser_t parsTime(char* oldTime)
{
    int32_t hours = atoi(strtok(oldTime, ":,"));
    int32_t minutes = atoi(strtok(NULL, ":,"));
    int32_t seconds = atoi(strtok(NULL, ":,"));
    int32_t millisecond = atoi(strtok(NULL, ":,"));

    TimeParser_t parser = {hours, minutes, seconds, millisecond};
    return parser;
}

TimeParser_t shiftTime(TimeParser_t* time, int32_t shiftBy)
{
    int32_t shiftedTime = getCollapsedTime(time) - shiftBy;
    if (shiftedTime < 0)
        return (TimeParser_t){0, 0, 0, 0};
        
    return (TimeParser_t){
        shiftedTime / 3600000 % 10000,
        shiftedTime % 3600000 / 60000,
        shiftedTime % 60000 / 1000,
        shiftedTime % 1000
    };
}

int8_t isTextNumber(const char* text)
{
    if (!*text)
        return 0;

    char* c = (char*)text;
    while (*c)
    {
        if (!isdigit(*c))
            return 0;
        c++;
    }
    return 1;
}

void shiftSubtitles(Timer_t* timer, int32_t shiftBy)
{
    shiftBy *= 1000;
    FILE* source = fopen(timer->source, "r");
    FILE* shiftedSubtitles = fopen(timer->shiftedSubtitles, "w");

    if (source == NULL || shiftedSubtitles == NULL)
    {
        puts("Could not find source or could not create a new file.");
        return;
    }

    uint32_t textNumber = 1;

    ssize_t textLen;
    size_t lineLen = 0;
    char* line = NULL;
    size_t hourLen = 0;
    char* hour = NULL;
    char* timeSplit;
        
    while ((textLen = getline(&line, &lineLen, source)) != -1)
    {
        line[textLen - 1] = '\0';

        if(!(isTextNumber(line)) && textNumber != 1) 
        {
            fprintf(shiftedSubtitles, "%s\n", line);
            continue;
        }

        textLen = getline(&hour, &hourLen, source);
        hour[textLen - 1] = '\0';

        timeSplit = strstr(hour, " --> ");
        *timeSplit = '\0';
        timeSplit += 5;

        TimeParser_t parsed1stPart = parsTime(hour);
        TimeParser_t shifted1stHalf = shiftTime(&parsed1stPart, shiftBy);

        TimeParser_t parsed2ndPart = parsTime(timeSplit);
        TimeParser_t shifted2ndHalf = shiftTime(&parsed2ndPart, shiftBy);

        fprintf(shiftedSubtitles, "%d\n%02d:%02d:%02d,%03d --> %02d:%02d:%02d,%03d\n",
            textNumber,
            shifted1stHalf.hours, shifted1stHalf.minutes, shifted1stHalf.seconds, shifted1stHalf.milliseconds,
            shifted2ndHalf.hours, shifted2ndHalf.minutes, shifted2ndHalf.seconds, shifted2ndHalf.milliseconds);

        textNumber += 1;
    }

    free(line);
    free(hour);
    fclose(source);
    fclose(shiftedSubtitles);
}

int main()
{
    Timer_t test = {"ST1.srt", "ST1S.srt"};

    shiftSubtitles(&test, 12);

    return 0;
}